specs/lightclient/sync-protocol.md

```diff
@@ -0,0 +1,132 @@
+- [Constants](#constants)
+- [Containers](#containers)
+  - [`LightClientUpdate`](#lightclientupdate)
+- [Helpers](#helpers)
+  - [`LightClientMemory`](#lightclientmemory)
+- [Light client state updates](#light-client-state-updates)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+Ethereum 2.0 is designed to be light client friendly. This allows low-resource clients such as mobile phones to access Ethereum 2.0 with reasonable safety and liveness. It also facilitates the development of "bridges" to external blockchains. This document suggests a minimal light client design for the beacon chain that uses the concept of "sync committees" introduced in [./beacon-chain.md](the the light-client-friendliness beacon chain extension).
+
+## Constants
+
+| Name | Value |
+| - | - |
+| `SYNC_COMMITTEES_GENERALIZED_INDEX` | `GeneralizedIndexConcat(GeneralizedIndex(BeaconBlock, 'state_root'), GeneralizedIndex(BeaconState, 'current_sync_committee'))` |
+| `FORK_GENERALIZED_INDEX` | `GeneralizedIndexConcat(GeneralizedIndex(BeaconBlock, 'state_root'), GeneralizedIndex(BeaconState, 'fork'))` |
+
+## Containers
+
+### `LightClientUpdate`
+
+```python
+class LightClientUpdate(Container):
+    # Updated beacon header (and authenticating branch)
+    header: BeaconBlockHeader
+    # Sync committee signature to that header
+    aggregation_bits: Bitlist[MAX_SYNC_COMMITTEE_SIZE]
+    signature: BLSSignature
+    header_branch: Vector[Bytes32, BEACON_CHAIN_ROOT_IN_SHARD_BLOCK_HEADER_DEPTH]
+    # Updates fork version
+    new_fork: Fork
+    fork_branch: Vector[Bytes32, log_2(FORK_GENERALIZED_INDEX)]
+    # Updated period committee (and authenticating branch)
+    new_current_sync_committee: SyncCommittee
+    new_next_sync_committee: SyncCommittee
+    sync_committee_branch: Vector[Bytes32, log_2(SYNC_COMMITTEES_GENERALIZED_INDEX)]
+```
+
+## Helpers
+
+### `LightClientMemory`
+
+```python
+class LightClientMemory(Container):
+    # Beacon header which is not expected to revert
+    header: BeaconBlockHeader
+    # Fork version data
+    fork_version: Version
+    # period committees corresponding to the beacon header
+    current_sync_committee: SyncCommittee
+    next_sync_committee: SyncCommittee
+```
+
+## Light client state updates
+
+The state of a light client is stored in a `memory` object of type `LightClientMemory`. To advance its state a light client requests an `update` object of type `LightClientUpdate` from the network by sending a request containing `(memory.shard, memory.header.slot, slot_range_end)`. It calls `validate_update(memory, update)` on each update that it receives in response. If `sum(update.aggregate_bits) * 3 > len(update.aggregate_bits) * 2` for any valid update, it accepts that update immediately; otherwise, it waits around for some time and then finally calls `update_memory(memory, update)` on the valid update with the highest `sum(update.aggregate_bits)`.
```

Comment:
> If `sum(update.aggregate_bits) * 3 > len(update.aggregate_bits) * 2` for any valid update

is there a reason we are not summing effective stake (via `effective_balance` found in `memory.current_sync_committee`?

Reply:
It vastly simplifies the light client spec to rely only on the bits. We can make up for it by making membership in the sync committee itself probabilistic.